command/agent/config_parse.go

```diff
@@ -30,1051 +27,173 @@ func ParseConfigFile(path string) (*Config, error) {
-	if o := listVal.Filter("reserved"); len(o.Items) > 0 {
-		if err := parseReserved(&config.Reserved, o); err != nil {
-			return multierror.Prefix(err, "reserved ->")
-		}
-	}
-
-	// Parse ServerJoin config
-	if o := listVal.Filter("server_join"); len(o.Items) > 0 {
-		if err := parseServerJoin(&config.ServerJoin, o); err != nil {
-			return multierror.Prefix(err, "server_join->")
-		}
-	}
-
-	*result = &config
-	return nil
-}
-
-func parseReserved(result **Resources, list *ast.ObjectList) error {
-	list = list.Elem()
-	if len(list.Items) > 1 {
-		return fmt.Errorf("only one 'reserved' block allowed")
-	}
-
-	// Get our reserved object
-	obj := list.Items[0]
-
-	// Value should be an object
-	var listVal *ast.ObjectList
-	if ot, ok := obj.Val.(*ast.ObjectType); ok {
-		listVal = ot.List
-	} else {
-		return fmt.Errorf("client value: should be an object")
-	}
-
-	// Check for invalid keys
-	valid := []string{
-		"cpu",
-		"memory",
-		"disk",
-		"reserved_ports",
-	}
-	if err := helper.CheckHCLKeys(listVal, valid); err != nil {
-		return err
-	}
-
-	var m map[string]interface{}
-	if err := hcl.DecodeObject(&m, listVal); err != nil {
-		return err
-	}
-
-	var reserved Resources
-	if err := mapstructure.WeakDecode(m, &reserved); err != nil {
-		return err
-	}
-	if err := reserved.CanParseReserved(); err != nil {
-		return err
-	}
-
-	*result = &reserved
-	return nil
-}
-
-func parseServer(result **ServerConfig, list *ast.ObjectList) error {
-	list = list.Elem()
-	if len(list.Items) > 1 {
-		return fmt.Errorf("only one 'server' block allowed")
-	}
-
-	// Get our server object
-	obj := list.Items[0]
-
-	// Value should be an object
-	var listVal *ast.ObjectList
-	if ot, ok := obj.Val.(*ast.ObjectType); ok {
-		listVal = ot.List
-	} else {
-		return fmt.Errorf("client value: should be an object")
-	}
-
-	// Check for invalid keys
-	valid := []string{
-		"enabled",
-		"bootstrap_expect",
-		"data_dir",
-		"protocol_version",
-		"raft_protocol",
-		"num_schedulers",
-		"enabled_schedulers",
-		"node_gc_threshold",
-		"eval_gc_threshold",
-		"job_gc_threshold",
-		"deployment_gc_threshold",
-		"heartbeat_grace",
-		"min_heartbeat_ttl",
-		"max_heartbeats_per_second",
-		"rejoin_after_leave",
-		"encrypt",
-		"authoritative_region",
-		"non_voting_server",
-		"redundancy_zone",
-		"upgrade_version",
-
-		"server_join",
-
-		// For backwards compatibility
-		"start_join",
-		"retry_join",
-		"retry_max",
-		"retry_interval",
-	}
-	if err := helper.CheckHCLKeys(listVal, valid); err != nil {
-		return err
-	}
-
-	var m map[string]interface{}
-	if err := hcl.DecodeObject(&m, listVal); err != nil {
-		return err
+		return nil, err
 	}
 
-	delete(m, "server_join")
-
-	var config ServerConfig
-	dec, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
-		DecodeHook:       mapstructure.StringToTimeDurationHookFunc(),
-		WeaklyTypedInput: true,
-		Result:           &config,
+	// convert strings to time.Durations
+	err = durations([]td{
```

Comment:
Could we add additional comments to make it clear that all duration fields will need the string equivalent from now on - probably best to add it in the config struct at the top

Reply:
@preetapan I added a comment right at the top of agent/config.go